agent/http.go

```diff
@@ -383,13 +383,43 @@ func parseWait(resp http.ResponseWriter, req *http.Request, b *structs.QueryOpti
 
 // parseConsistency is used to parse the ?stale and ?consistent query params.
 // Returns true on error
-func parseConsistency(resp http.ResponseWriter, req *http.Request, b *structs.QueryOptions) bool {
+func (s *HTTPServer) parseConsistency(resp http.ResponseWriter, req *http.Request, b *structs.QueryOptions) bool {
 	query := req.URL.Query()
+	defaults := true
 	if _, ok := query["stale"]; ok {
 		b.AllowStale = true
+		defaults = false
 	}
 	if _, ok := query["consistent"]; ok {
 		b.RequireConsistent = true
+		defaults = false
+	}
+	if _, ok := query["leader"]; ok {
+		defaults = false
+	}
+	if maxStale := query.Get("max_stale"); maxStale != "" {
+		fmt.Printf("**** max_stale=%s\n", maxStale)
+		dur, err := time.ParseDuration(maxStale)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
+			fmt.Fprint(resp, "Invalid max_stale time")
+			return true
+		}
+		b.MaxStaleDuration = dur
+		if dur.Nanoseconds() > 0 {
+			b.AllowStale = true
+			defaults = false
+		}
+	}
+	// No specific Consistency has been specified by caller
+	if defaults {
+		path := req.URL.Path
+		if strings.HasPrefix(path, "/v1/catalog") || strings.HasPrefix(path, "/v1/health") {
+			b.MaxStaleDuration = s.agent.config.DiscoveryMaxStale
+			if b.MaxStaleDuration.Nanoseconds() > 0 {
+				b.AllowStale = true
+			}
+		}
```

Comment:
Upon further discussion on this, we thought it would be nice to add a header when the default consistency mode is overridden, so one suggestion we have is to add a new response header. This method could add a header like `resp.AddHeader("X-Consul-Effective-Consistency-Level","stale")` or `resp.AddHeader("X-Consul-Effective-Consistency-Level","leader")` etc. That way if a client lib is making a default request to an agent that has DiscoveryMaxStale set and is being overridden by this block, the response should tell you what it used.

Reply:
Ok, doing it as well